### ISSUE TITLE
useState of the Tuple type is not safe

### ISSUE THREAD

**Issue by Kasik-D** [reactjs/reactjs.org#5940]
I'm reading TS and noticed a vulnerability like UseState, as described in the documentation, we get a tuple of two elements. 

The tuple in TS has a vulnerability because you can use array methods like pop, push on them. Which isn't safe because TS doesn't compute the type after that, it stays. 

I suggest adding a readonly modifier to avoid these situations 

```typescript
  const state = useState();

  state.pop();

  const [value, setValue] = state;
  
  // setValue is undefined, but ts shows that the value is of type React.Dispatch<React.SetStateAction<undefined>
```

**Comment by MeenuyD** [comment]
Yes, you are correct that using array methods like pop and push on the returned tuple of useState in TypeScript can cause issues with the inferred types. This is because TypeScript is not able to compute the updated type after these methods are called.

To avoid this issue, you can use the read-only modifier to prevent modifying the returned tuple directly. Instead, you can destructure the tuple and use the returned variables to modify the state:
```
const [value, setValue] = useState();

// This will result in a TypeScript error because the state is read-only
state.pop();

// Use the destructured variables to modify the state
setValue(newValue);

```
By using the destructured variables, TypeScript can infer the correct types and avoid any issues with modifying the returned tuple directly.
